Updates to address Martin Duke's DISCUSS (#226)
tlswg/dtls13-spec: draft-ietf-tls-dtls13.md

### Timer Values

Unless they have extra information about the
round trip time, implementations SHOULD use an initial timer value
of 1000 ms and double the value at each retransmission, up to no less than 60 seconds
(the RFC 6298 {{RFC6298}} maximum). Application specific profiles, MAY recommend shorter
or longer timer values. For instance:

> * Restricted profiles such as IoT MAY specify longer timeouts.

Reviewer: ```suggestion
* Profiles for specific deployment environments, such as in low-power, multi-hop mesh network as used in some Internet of Things (IoT) networks, MAY specify longer timeouts. See {{?I-D.ietf-uta-tls13-iot-profile}} for more information about one such DTLS 1.3 IoT profile.
```

* Real-time protocols MAY specify shorter timeouts. It is RECOMMENDED
  that for DTLS-SRTP {{?RFC5764}}, a default timeout of
  400ms be used; because customer experience degrades with one-way latencies
  of greater than 200ms, real-time deployments are less likely
  to have long latencies. 

In settings where external information (for instance from an ICE {{?RFC8445}}
handshake, or from previous connections to the same server)
about the RTT, implementations SHOULD use that to set the retransmit
time. 

Implementations SHOULD retain the current timer value until a
message is transmitted and acknowledged without having to
be retransmitted, at which time the value SHOULD be reset
to approximately 1.5 times the measured round trip time for that
message. After a long period of idleness, no less
than 10 times the current timer value, implementations MAY reset the
timer to the initial value.

Note that because retransmission for handshake and not dataflow, the effect on
congestion of shorter timeouts is smaller than in generic protocols
such as TCP or QUIC. Experience with DTLS 1.2, which uses a

> simpler "retransmit everything on timeout" approach has not shown

Reviewer: ```suggestion
simpler "retransmit everything on timeout" approach, has not shown
```

deployment environments require timer value adjustments. Mishandling
of the timer can lead to serious congestion problems, for example if
many instances of a DTLS time out early and retransmit too quickly on
a congested link.

> Unless implementations have extra information about the round trip time, 

Reviewer 2: ```suggestion
Unless implementations have deployment-specific and/or external information about the round trip time, 
```

implementations SHOULD use an initial timer value of 1000 ms and double 
the value at each retransmission, up to no less than 60 seconds (the 

> RFC 6298 {{RFC6298}} maximum). Application specific profiles, MAY 

Reviewer 2: ```suggestion
{{RFC6298}} maximum). Application-specific profiles, MAY 
```

recommend shorter or longer timer values. For instance:

* Profiles for specific deployment environments, such as in low-power, 
  multi-hop mesh network as used in some Internet of Things (IoT) networks, 
  MAY specify longer timeouts. See {{?I-D.ietf-uta-tls13-iot-profile}} for 
  more information about one such DTLS 1.3 IoT profile.

* Real-time protocols MAY specify shorter timeouts. It is RECOMMENDED
  that for DTLS-SRTP {{?RFC5764}}, a default timeout of
  400ms be used; because customer experience degrades with one-way latencies
  of greater than 200ms, real-time deployments are less likely
  to have long latencies. 

> In settings where external information (for instance from an ICE {{?RFC8445}}

Reviewer 3: ```suggestion
In settings where there is external information (for instance from an ICE {{?RFC8445}}
```

handshake, or from previous connections to the same server)
about the RTT, implementations SHOULD use that to set the retransmit

> time. 

Reviewer 2: There should probably be some text about how to "use that to set the retransmit time", i.e., simply reusing that value is not what TCP does. I think Linux does something like `RTO = max(SRTT + 4 * RTTVAR, 1000ms)`. [QUIC does something similar](https://www.ietf.org/archive/id/draft-ietf-quic-recovery-34.html#section-6.2.1).

Author: Sure. I'll make this 1.5RTT as suggested by Allman and used elsewhere in this doc,

Implementations SHOULD retain the current timer value until a
message is transmitted and acknowledged without having to
be retransmitted, at which time the value SHOULD be reset

> to approximately 1.5 times the measured round trip time for that

Reviewer 2: Why "approximately"?

Author: All of these values are approximate because timers are approximate.

message. After a long period of idleness, no less
than 10 times the current timer value, implementations MAY reset the
timer to the initial value.

> Note that because retransmission for handshake and not dataflow, the effect on

Reviewer 3: ```suggestion
Note that because retransmission is for the handshake and not dataflow, the effect on
```

messages are acknowledged. DTLS offers a number of mechanisms
for minimizing the size of the certificate message, including
the cached information extension {{?RFC7924}} and certificate

> compression {{?RFC8879}}. DTLS stacks MUST NOT send flights

Reviewer 4: The mailing list discussion recommended a SHOULD NOT, as far as I understand. It's not clear to me how a MUST NOT could be enforced in practice, if the e.g. the certificate chain is very long.

The issue here might be an overloading of the term "flight": As I read the paragraph, you use it not in the sense defined (https://www.ietf.org/archive/id/draft-ietf-tls-dtls13-41.html#name-dtls-handshake-flights) but in the sense of the part of the current flight which is currenly attempted to be transmitted. Using this latter definition, a MUST NOT as above could indeed be enforced, but ten then I think a different term should be used.

Author: Agreed. Fixed.
